### PR TITLE
feat: no longer need to check in `swift_deps_index.json` file and other big updates

### PR DESCRIPTION
**Pull request by cgrindel**
Lots of big changes in this PR:

- The SPM-related rules no longer depend upon a Swift deps index JSON file.
- Remove the `from_file` module extension tag class, as the ruleset no longer needs to read the Swift deps index file.
- Remove the `swift_update_packages` macro as it is no longer required.
- Remove support for `byName` references outside of a Swift package. This was supported in early SPM versions. It was one of the driving factors for having to pre-process the Swift dependencies and generate a Swift deps index file. Removing this support reduces the complexity of the `rules_swift_package_manager` implementation.
- Remove `update_repos` command from the Gazelle plugin.
- Add `swift_deps_info` repository rule to generate a Swift deps index JSON file as part of the build, not in the Gazelle plugin. This index file is used by the Gazelle plugin to resolve Swift module names to their respective Bazel target.
- Introduce `from_package` module extension tag class telling the ruleset where the `Package.swift` and `Package.resolved` files for the workspace live. These are used to identify the Swift package transitive dependencies for the project.
- Remove the `http_archive_ext_deps` example. The example existed to demonstrate and test indexing of Swift targets that were downloaded using `http_archive`. The Gazelle internals no longer provide some of the information needed for this to work properly with bzlmod enabled. Given that SPM has been widely adopted and there are some workarounds, we are removing support for this functionality in the Gazelle plugin when bzlmod is enabled.
- Add output to the example test runner that demarcates different phases for the testing.

Related to #924.

### CODE REVIEW

File: swiftpkg/bzlmod/swift_deps.bzl [hunk 1]
```diff
@@ -1,39 +1,154 @@
...
+        pkg_info_label = "@{}//:pkg_info.json".format(bazel_repo_name)
+        direct_dep_pkg_infos[pkg_info_label] = dep.identity
+
+    # Write info about the Swift deps that may be used by external tooling.
+    swift_deps_info_repo_name = "swift_deps_info"
+    swift_deps_info(
+        name = swift_deps_info_repo_name,
+        direct_dep_pkg_infos = direct_dep_pkg_infos,
+    )
+    direct_dep_repo_names.append(swift_deps_info_repo_name)
```
Comment (brentleyjones) [inline swiftpkg/bzlmod/swift_deps.bzl:61]: This should be done in a different extension, so projects that don't use it won't incur the cost (or the `use_repo` warning).

File: swiftpkg/bzlmod/swift_deps.bzl [hunk 2]
```diff
@@ -1,39 +1,154 @@
...
+            break
+        processing = to_process
+        to_process = []
+        for dep in processing:
+            dep_pkg_info = pkginfos.get(
+                module_ctx,
+                directory = dep.file_system.path,
+                debug_path = None,
+                resolved_pkg_map = None,
+                # repo_name = bazel_repo_names.from_identity(dep.identity),
```
Comment (brentleyjones) [inline swiftpkg/bzlmod/swift_deps.bzl:100]: Comment

File: swiftpkg/bzlmod/swift_deps.bzl [hunk 3]
```diff
@@ -1,39 +1,154 @@
...
+    # A local Swift package can reference other local Swift packages. Hence, we
+    # need to check all of the transitive local Swift packages, not just the
+    # direct local packages. We do not need to worry about the source control
+    # deps because they are already listed in the Package.resolved.
+    to_process = [
+        dep
+        for dep in all_deps_by_id.values()
+        if dep.file_system
+    ]
+    for _ in range(100):
```
Comment (brentleyjones) [inline swiftpkg/bzlmod/swift_deps.bzl:89]: Since this is supposed to represent a `do ... while` loop, maybe pull this out to a global constant, and make the range larger? Also, check if `to_process` is non-empty after the loop and throw an error?

File: swiftpkg/bzlmod/swift_deps.bzl [hunk 4]
```diff
@@ -1,39 +1,154 @@
...
+        if len(to_process) == 0:
```
Comment (brentleyjones) [inline swiftpkg/bzlmod/swift_deps.bzl:90]: I believe idiomatic Starlark is this:
```suggestion
        if not to_process:
```

File: swiftpkg/internal/repo_rules.bzl [hunk 5]
```diff
@@ -70,8 +70,8 @@ def _gen_build_files(repository_ctx, pkg_ctx):
     for target in pkg_info.targets:
         # Unfortunately, Package.resolved does not contain test-only external
         # dependencies. So, we need to skip generating test targets. If a target
-        # does not have any product memberships, it is a testonly
-        if target.type == "test" or len(target.product_memberships) == 0:
+        # does not have any product memberships, it is a testonly.
```
Comment (brentleyjones) [inline swiftpkg/internal/repo_rules.bzl:73]: Since we no longer check `target.product_memberships`, does this comment still hold?

File: swiftpkg/tests/pkginfo_target_deps_tests.bzl [hunk 6]
```diff
@@ -19,37 +18,96 @@ pkginfo_target_deps = make_pkginfo_target_deps(
...
-    target_name = "Foo",
+_internal_target_ref_with_condition = pkginfos.new_target_reference(
+    target_name = _foo_target.name,
     condition = _target_dep_condition,
 )
+
+# NOTE: Not sure if we need to support this. May products have a target
+# with the same name. We prioritize finding the target.
+# _internal_product_by_name = pkginfos.new_by_name_reference(_bar_product.name)
+
```
Comment (brentleyjones) [inline swiftpkg/tests/pkginfo_target_deps_tests.bzl:106]: Not needed?

File: swiftpkg/tests/pkginfo_target_deps_tests.bzl [hunk 7]
```diff
@@ -211,40 +205,54 @@ def _bzl_select_list_test(ctx):
...
+        #         bzl_selects.new(
+        #             kind = pkginfo_target_deps.target_dep_kind,
+        #             value = [
+        #                 bazel_labels.normalize(
+        #                     "@swiftpkg_mypackage//:Bar",
+        #                 ),
+        #             ],
+        #         ),
+        #     ],
+        # ),
```
Comment (brentleyjones) [inline swiftpkg/tests/pkginfo_target_deps_tests.bzl:238]: Not needed

**Review by brentleyjones — COMMENTED** [review]
Just some minor comments. I tested it locally and it works with our project!

File: swiftpkg/bzlmod/swift_deps.bzl (hunk 1 above)
Comment (cgrindel) [reply swiftpkg/bzlmod/swift_deps.bzl]: I spent some time trying to move this to a separate extension and not have to manually enter the Swift package identity to pkg_info.json label mapping. It turns out to be non-trivial. For now, I have added a `bool` attribute that controls whether the `swift_deps_info` is declared. The default is `False`.

**Comment by cgrindel** [comment]
@brentleyjones @luispadron This should be ready for another review.

File: swiftpkg/bzlmod/swift_deps.bzl [hunk 8]
```diff
@@ -42,46 +161,61 @@ def _declare_pkg_from_package(package, deps_index_label, config_pkg):
...
+        ),
+        "swift": attr.label(
             mandatory = True,
-            doc = "A `swift_deps_index.json`.",
+            allow_files = [".swift"],
+            doc = "A `Package.swift`.",
         ),
     },
-    doc = "Load Swift packages from a file generated by the Gazelle extension.",
+    doc = "Load Swift packages from a `Package.swift` and `Package.resolved`.",
```
Comment (brentleyjones) [inline swiftpkg/bzlmod/swift_deps.bzl:214]:
```suggestion
    doc = "Load Swift packages from `Package.swift` and `Package.resolved` files.",
```

File: swiftpkg/bzlmod/swift_deps.bzl [hunk 9]
```diff
@@ -42,46 +161,61 @@ def _declare_pkg_from_package(package, deps_index_label, config_pkg):
...
 )
 
 _configure_package_tag = tag_class(
-    attrs = {
+    attrs = dicts.add({
```
Comment (brentleyjones) [inline swiftpkg/bzlmod/swift_deps.bzl:218]: You can use the `dict` union operator now instead: `|`

File: swiftpkg/bzlmod/swift_deps.bzl [hunk 10]
```diff
@@ -91,17 +225,17 @@ _configure_package_tag = tag_class(
             mandatory = True,
         ),
         "recursive_init_submodules": attr.bool(
-            default = True,
+            default = False,
```
Comment (brentleyjones) [inline swiftpkg/bzlmod/swift_deps.bzl:228]: Why the change?

Comment (cgrindel) [reply swiftpkg/bzlmod/swift_deps.bzl:228]: This was an accident. Must have happened during a rebase.

**Review by brentleyjones — APPROVED** [review]

**Comment by brentleyjones** [comment]
@luispadron can you give this a spin locally as well?

**Comment by luispadron** [comment]
Yes, I'll test and review today!

**Review by luispadron — APPROVED** [review]
